### PR TITLE
Finish adding gdk::Atom support to GTK

### PR DESCRIPTION
**Pull request by Susurrus**
Still don't know why this type is necessary, but realized I needed to add it after needing it for #472.

### CODE REVIEW

File: src/auto/accel_group.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// This file was generated by gir (b7f5189) from gir-files (71d73f0)
+// This file was generated by gir (b7f5189) from gir-files (71d73f0+)
```
Comment (EPashkin) [inline src/auto/accel_group.rs]: Please, repeat Regen without +

**Comment by Susurrus** [comment]
I believe this requires additional changes to `gdk` for the missing `FromGlibPtr*` impls I think, which should be trivial.

**Comment by EPashkin** [comment]
Seems It was bad idea removing `impl FromGlibPtrNone<ffi::GdkAtom> for Atom {` :(

File: src/auto/accel_group.rs (hunk 1 above)
Comment (Susurrus) [inline src/auto/accel_group.rs]: I don't know what that `+` is from as my working directory is clean. Ran submodule update and the submodule is pointed to the correct git version. Manually checked out master and pulled to the latest version and it's still giving me that plus. You ever see that anywhere?

Comment (EPashkin) [inline src/auto/accel_group.rs]: Remove gir.exe from target. Maybe it detect that no changes and not rebuild `gir_version.rs`

Comment (EPashkin) [inline src/auto/accel_group.rs]: Or do some change and reset it

Comment (Susurrus) [inline src/auto/accel_group.rs]: I deleted it and rechecked it out, tried deleting `gir-files/target` and rebuilt everything, nothing I have is really doing the trick here. I've seen this show `-dirty`, but never a `+` so I don't even know what that means.

File: src/auto/text_buffer.rs [hunk 2]
```diff
@@ -274,9 +284,13 @@ impl TextBuffer {
         }
     }
 
-    //pub fn get_serialize_formats(&self) -> (/*Ignored*/Vec<gdk::Atom>, i32) {
-    //    unsafe { TODO: call ffi::gtk_text_buffer_get_serialize_formats() }
-    //}
+    pub fn get_serialize_formats(&self) -> (Vec<gdk::Atom>, i32) {
```
Comment (EPashkin) [inline src/auto/text_buffer.rs]: This function need ignored or implemented manually.
Test shows 1/6 and fails on printing second name.
```
let buffer = gtk::TextBuffer::new(None);
buffer.set_text("test");

let (formats, count) = buffer.get_serialize_formats();
println!("{}/{}", count, formats.len());
for format in formats {
    println!("{}", format.name());
}
```

File: src/auto/text_buffer.rs [hunk 3]
```diff
@@ -165,9 +171,13 @@ impl TextBuffer {
     //    unsafe { TODO: call ffi::gtk_text_buffer_get_copy_target_list() }
     //}
 
-    //pub fn get_deserialize_formats(&self) -> (/*Ignored*/Vec<gdk::Atom>, i32) {
-    //    unsafe { TODO: call ffi::gtk_text_buffer_get_deserialize_formats() }
-    //}
+    pub fn get_deserialize_formats(&self) -> (Vec<gdk::Atom>, i32) {
```
Comment (EPashkin) [inline src/auto/text_buffer.rs]: Same as `get_serialize_formats` not sure how to get it return something

File: src/auto/text_buffer.rs (hunk 2 above)
Comment (EPashkin) [inline src/auto/text_buffer.rs]: Updated test. Seems need use `from_glib_container_num` instead `from_glib_container`. Not checked this.

File: src/auto/accel_group.rs (hunk 1 above)
Comment (EPashkin) [inline src/auto/accel_group.rs]: `+` after hash mean that you `gir` dir not clean for git.

File: src/auto/text_buffer.rs (hunk 3 above)
Comment (EPashkin) [inline src/auto/text_buffer.rs]: It can return something after `register_deserialize_tagset`

File: src/auto/accel_group.rs (hunk 1 above)
Comment (EPashkin) [inline src/auto/accel_group.rs]: Hash saved to `src/gir_version.rs` but sometimes build silently fails to do that (or think that nothing changed or just don't run `build.rs`)

Comment (EPashkin) [inline src/auto/accel_group.rs]: Hm, I missed that + in gir-files version.
Check `git status` in `gir-files` directory: it must say `nothing to commit, working tree clean`
After this remove `gir/target` and try rebuild/

File: src/auto/selection_data.rs [hunk 4]
```diff
@@ -21,9 +21,11 @@ impl SelectionData {
     //    unsafe { TODO: call ffi::gtk_selection_data_get_data() }
     //}
 
-    //pub fn get_data_type(&self) -> /*Ignored*/Option<gdk::Atom> {
-    //    unsafe { TODO: call ffi::gtk_selection_data_get_data_type() }
-    //}
+    pub fn get_data_type(&self) -> Option<gdk::Atom> {
```
Comment (EPashkin) [inline src/auto/selection_data.rs]: @Susurrus please test any of these functions to be sure that `from_glib_none` works on atom.

Comment (EPashkin) [inline src/auto/selection_data.rs]: Sorry, tested it on results `register_serialize_tagset` all works.

**Comment by EPashkin** [comment]
Also seems [register_serialize_tagset](https://developer.gnome.org/gtk3/stable/GtkTextBuffer.html#gtk-text-buffer-register-serialize-tagset), register_deserialize_tagset, [get_target](https://developer.gnome.org/gtk3/stable/gtk3-Selections.html#gtk-selection-data-get-target), get-selection and get_data_type has non-optional return

File: src/auto/text_buffer.rs (hunk 2 above)
Comment (Susurrus) [inline src/auto/text_buffer.rs]: I don't quite understand this, but it sounds like I should set the function to ignore during generation.

Comment (EPashkin) [inline src/auto/text_buffer.rs]: Better both ignore and add it as manual with other conversion

**Comment by Susurrus** [comment]
I've updated things and now I think we're closer. For those functions erronously returning an `Option<_>` type, I just ignored them. Those should be addressed by changes to `gir` or `gir-files`.

**Comment by EPashkin** [comment]
I mean not ignoring non-optional functions but explain it to gir like in https://github.com/gtk-rs/gtk/blob/master/Gir.toml#L325-L326

**Comment by Susurrus** [comment]
Thanks! I figured there had to be a way to set things like this in the `Gir.toml`. Here's the latest, but I didn't deal with `Clipboard:get_(de)serialize_formats()` yet (just set them as ignored) as I'm not positive what they should return. From reading them I think they should just return a `Vec<gdk::Atom>` as the length doesn't need to be returned as a separate object. If that's the case, then I can implement it manually or should this be something done by `gir`?

**Comment by EPashkin** [comment]
Currently `Clipboard:get_(de)serialize_formats()` need manual implementation. Agree that it need return just vector.

**Comment by Susurrus** [comment]
Alright, I think that resolves all issues you've brough up @EPashkin. Note that CI won't succeed until gtk-rs/gdk#148 is merged.

File: src/text_buffer.rs [hunk 5]
```diff
@@ -4,16 +4,31 @@
...
 use glib::signal::connect;
 use glib_ffi;
 use TextBuffer;
 use TextIter;
 
 impl TextBuffer {
+    pub fn get_deserialize_formats(&self) -> Vec<gdk::Atom> {
+        unsafe {
+            let mut n_formats = mem::uninitialized();
+            FromGlibPtrContainer::from_glib_container(ffi::gtk_text_buffer_get_deserialize_formats(self.to_glib_none().0, &mut n_formats))
```
Comment (EPashkin) [inline src/text_buffer.rs]: As I says it need be `from_glib_container_num`

Comment (Susurrus) [inline src/text_buffer.rs]: Then what should the `num` argument be when calling it? The total number is returned by the type not specified by the caller, which is why I left it as is. But the docs of `from_glib_container_num` are sparse.

Comment (EPashkin) [inline src/text_buffer.rs]: It number of elements in array. In this case `n_formats`

**Comment by Susurrus** [comment]
Sorry, @EPashkin, looks like I was a little too tired, I understand your point about `get_(de)serialize_formats()`. This should be ready to roll now!

**Comment by EPashkin** [comment]
@Susurrus, Thanks, It seems good.
I check with https://github.com/gtk-rs/gtk/pull/478#discussion_r108222841 just in case in 3 hour.
Then we need wait until @GuillaumeGomez have free time, merged https://github.com/gtk-rs/gdk/pull/148 and restart CI.

**Comment by GuillaumeGomez** [comment]
@EPashkin: gtk-rs/gdk#148 needs some modifications so we'll have to wait a bit. ;)

**Comment by EPashkin** [comment]
Strange but `get_deserialize_formats` still return too big vector that panic on iterate.

**Comment by GuillaumeGomez** [comment]
Restarting CI.

**Comment by EPashkin** [comment]
Very strange implementation in https://github.com/gtk-rs/glib/blob/master/src/translate.rs#L886-L937
for *_num functions. `Num` used only for capacity.
Seems this need fix too.

**Comment by EPashkin** [comment]
And others too 😭

**Comment by EPashkin** [comment]
All CI finished. glib path don't change signatures so can't fail build.

**Comment by GuillaumeGomez** [comment]
Great, then I merge. Thanks @Susurrus!

**Pull request merged**